### PR TITLE
The TheiaValidate Workflow

### PR DESCRIPTION
**Pull request by sage-wright**
This PR adds the TheiaValidate workflow to PHB.

## :hammer_and_wrench: Changes Being Made

- The TheiaValidate_PHB workflow has been created

## :brain: Context and Rationale

This workflow performs basic comparisons between user-designated columns in two separate tables. We anticipate this workflow being run to determine the differences, if any, between version releases, or differences in results between two workflows, such as TheiaProk_ONT vs TheiaProk_Illumina_PE.

In order to enable this workflow to work on different workflow series, we have required users to provide a list of columns they want to compare between two tables. If additional validation metrics are desired, the user has the ability to provide a validation_criteria_tsv file that specifies what type of comparison should be performed.

## :clipboard: Workflow/Task Steps

1. Export the two different tables from Terra
2. If the two tables are the same length (same _number_ of samples), attempt validation. This does _not_ check if the samples have the same names. _Validation will not work if the names of samples are not the same between the two tables_.
3. Count for each column requested the number of samples that have data present. For those columns that have data present, count how many match exactly (string equivalence).
4. Create a table that compares the two data frames showing only the content when there are differences between the two tables (matches are blank) and writes it to an Excel file.
5. If a validation_criteria_tsv file was provided, additional validation checks will occur. Right now, there are several options:
   - **EXACT** performs an exact string match and counts the number of exact match failures/differences
   - **IGNORE** does not check the values, and says there are 0 failures
   - **SET** checks list items (such as `amrfinder_plus_genes`) for identical content -- order does not matter.
   - **FLOAT** which is an actual float, such as **0.02**. calculates the percent difference between _numerical_ columns. If the columns are not numerical, this function will not work. Example: If the float is 0.02, the test will indicate a failure if the values in the columns are more than 2% different. 
  Date values, integers, and object-type values are ignored and indicate 0 failures.
6. A PDF is created that displays the summary information generated in Steps 2, 3, and 5 (if provided).

### Inputs
 **_Mandatory_**:
- _table1_: the name of the first table
- _table2_: the name of the second table
- _columns_to_compare_: a comma-separated list of columns to compare between the two tables
- _output_prefix_: the name of the output files
- _terra_workspace1_name_: the name of the Terra workspace where table1 can be found
- _terra_project1_name_: the name of the Terra project where table1 can be found

**_Optional_**:
- _terra_workspace2_name_: the name of the Terra workspace where table2 can be found
- _terra_project2_name_: the name of the Terra project where table2 can be found
- _validation_criteria_tsv_: a TSV file that lists the additional criteria to check. This file takes the following format (a header line is required): 
 
```
column	criteria
amrfinderplus_amr_genes	SET
assembly_length	0.01
gambit_predicted_taxon	EXACT
plasmidfinder_plasmids	SET
seqsero2_predicted_antigenic_profile	EXACT
serotypefinder_serotype	EXACT
ts_mlst_predicted_st	EXACT
```

### Outputs

- _theiavalidate_version_: the version of the PHB repository
- _theiavalidate_date_: the date the analysis was run
- _validation_report_: the PDF summary report
- _validation_differences_table_: the Excel spreadsheet listing only differences
- _validation_status_: Indicates whether or not validation was attempted.

Example PDF report:
output_PDF

Example Excel file:
[test.xlsx](https://github.com/theiagen/public_health_bioinformatics/files/11638570/test.xlsx)

### Locally

Cannot test the entire workflow locally due to permissions, but the `compare_two_tsvs` task has been tested with successful results.

### Terra

Successful job here: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/8ba8494e-7477-4f56-8ea8-eb9c7ffe746b
## :microscope: Quality checks

Pull Request (PR) checklist:
- [x] Include a description of what is in this pull request in this message.
- [x] The workflow/task has been tested locally and on Terra
- [x] The CI/CD has been adjusted and tests are passing
- [x] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

File: tasks/utilities/task_validate.wdl [hunk 1]
```diff
@@ -2,54 +2,284 @@ version 1.0
...
-    String? out_prefix
+    File? validation_criteria_tsv
+    String columns_to_compare
+    String output_prefix
+
     Int disk_size = 10
   }
   command <<<
-    compare-data-tables.py ~{datatable1_tsv} ~{datatable2_tsv} --outdir ~{out_dir} --prefix ~{out_prefix}
+    # too lazy to create a new docker image, this is not good practice
```
Comment (kevinlibuit) [inline tasks/utilities/task_validate.wdl:66]: Hah! I literally just wrote this out before seeing this comment: 

> python block is getting a bit complex; we should consider storing this into a separate repo and building a docker for these types of utilities

Comment (sage-wright) [reply tasks/utilities/task_validate.wdl:66]: My one problem with making it into a script included in a docker is that whenever we want to make updates to the code, we'll have to make a new docker image as well. That's not a deal-breaker, but it's an extra step and an extra level of obscurity when people have to figure out where to find the code that performs the comparisons. 

A large part of this python block is just comments explaining what's going on as well

**Comment by michellescribner** [comment]
@sage-wright This looks great, I'm launching a test now! The docs that you wrote look awesome as well. I'd suggest adding a blurb explaining that the user should select "Run workflow with inputs defined by file paths" rather than a root entity type.

**Comment by michellescribner** [comment]
Launched test of Taxon_Full_Illumina_PE_TEMPLATE table from PHBG v1.1.0 validation versus mrs-amrfinder-scope-dev analysis on the same samples. This dev branch split the amrfinderplus_amr_genes column into amrfinderplus_amr_core_genes and amrfinderplus_amr_plus_genes. We should see that the amrfinderplus_amr_genes column is not present in the second data table. 
https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/b6a410a2-1d31-4b97-8de2-b79bfb4ff6ae

Columns compared: `"ani_highest_percent,ani_highest_percent_bases_aligned,assembly_length,busco_results,combined_mean_q_clean,combined_mean_q_raw,combined_mean_readlength_clean,combined_mean_readlength_raw,est_coverage_clean,est_coverage_raw,gambit_predicted_taxon,midas_secondary_genus_abundance,n50_value,number_contigs,r1_mean_q,r1_mean_q_clean,r1_mean_q_raw,r1_mean_readlength_raw,r2_mean_q,r2_mean_q_clean,r2_mean_q_raw,r2_mean_readlength_raw,amrfinderplus_amr_genes,amrfinderplus_amr_core_genes,amrfinderplus_amr_plus_genes"`

Validation criteria:
```
column	criteria
amrfinderplus_amr_genes	SET
assembly_length	0.01
gambit_predicted_taxon	EXACT
plasmidfinder_plasmids	SET
seqsero2_predicted_antigenic_profile	EXACT
serotypefinder_serotype	EXACT
ts_mlst_predicted_st	EXACT
```

**Comment by michellescribner** [comment]
Successful test: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/d01dfe77-4efa-49d4-b813-94e43fcc2409

This workflow is ~chef's kiss~ once it passes GHA

**Comment by sage-wright** [comment]
I'm honestly not sure how to fix the GHA because the miniwdl check task is being triggered to check a file after changes are made, but the change made is the file being removed, which makes the miniwdl check fail because the file doesn't exist. 🤷

**Review by michellescribner — APPROVED** [review]
Celebrate good times, come on

**Pull request merged**